src/hunk_index.rs

```diff
@@ -0,0 +1,133 @@
+use std::cmp::Ordering;
+
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+
+use crate::{Apath, IndexRead, Result};
+
+#[derive(Debug)]
+struct HunkIndexMeta {
+    index: u32,
+
+    start_path: Apath,
+    end_path: Apath,
+}
+
+/// An index over all available hunks available in an index
+/// for speeding up sub-dir iterations and locating
+/// path metadata.
+pub struct IndexHunkIndex {
```

Comment:
Maybe this should merge with `IndexRead`, though not necessarily in this PR.

Reply:
I think `IndexRead` and `IndexHunkIndex` do have two different purposes.
`IndexRead`'s purpose is to read information from an archive. In contrast, `IndexHunkIndex` builds an optional, but not necessary, lookup table for hunks, which can be used to make `IndexRead` more efficient. Hence, I reason that `IndexRead` and `IndexHunkIndex` serve different purposes and should be kept separate.